### PR TITLE
Dotnet test preserve results

### PR DESCRIPTION
**Pull request by dillon-ms**
Test Results under the circle ci Tests tab
failure:
Screenshot 2022-10-05 at 17-00-34 build_test (159) - butlfrazp_CircleCITest
success:
image

Coverage Report under Artifacts -> index.html
image

- add these pkgs to your test proj: `JunitXml.TestLogger` and `Microsoft.NET.Test.Sdk`
- call test like this: 
```
dotnet test CircleCiTestFunction/CircleCiTestFunction.sln 
    /p:CollectCoverage=true 
    /p:CoverletOutputFormat=cobertura 
   /p:Threshold=<< parameters.threshold >>
    --logger:"junit;LogFilePath=TestResults/NUnit.xml"
```

If the coverage threshold is not met, the coverage reports will still be generated and uploaded, then an error will be thrown that will mark the pipeline as failed.

### CODE REVIEW

File: .circleci/continue_config.yml [hunk 1]
```diff
@@ -50,18 +50,37 @@ commands:
...
           name: Build solution 
           command: dotnet build CircleCiTestFunction/CircleCiTestFunction.sln
       - run:
-          name: "Running tests"
-          command: dotnet test CircleCiTestFunction/CircleCiTestFunction.sln
-
+          name: Run tests and build coverage report
+          command: |
+            # build reports even if dotnet test fails, then throw to inform Circle CI to mark the step as failed
+            dotnet test CircleCiTestFunction/CircleCiTestFunction.sln /p:Threshold=<< parameters.threshold >> /p:CollectCoverage=true /p:CoverletOutputFormat=cobertura --logger:"junit;LogFilePath=TestResults/NUnit.xml"
```
Comment (nairmai) [inline .circleci/continue_config.yml]: just noting that we may or may not be using NUnit as the test framework - I think Mahdi had mentioned we might prefer XUnit, might be worth syncing/aligning on whatever that decision is

Comment (dillon-ms) [reply .circleci/continue_config.yml]: Sure I'll check with him

File: .circleci/continue_config.yml [hunk 2]
```diff
@@ -61,17 +61,66 @@ commands:
...
+        type: integer
     steps:
       - checkout
+      - run:
+          name: Get report generator
+          command: dotnet tool install dotnet-reportgenerator-globaltool --version 5.1.10 --tool-path Tools/
       - run:
           name: Restore solution dependencies and tools
-          command: dotnet restore CircleCiTestFunction/CircleCiTestFunction.sln
+          command: dir "*.csproj" -Recurse | %{dotnet restore $_.FullName}
```
Comment (msetbar) [inline .circleci/continue_config.yml:75]: dotnet restore solution file does not work?

File: .circleci/continue_config.yml [hunk 3]
```diff
@@ -61,17 +61,66 @@ commands:
...
+          command: dir "*.csproj" -Recurse | %{dotnet build $_.FullName}
       - run:
-          name: "Running tests"
-          command: dotnet test CircleCiTestFunction/CircleCiTestFunction.sln
+          name: Run tests and build coverage report
+          command: |
+
+            # run dotnet test on all projects ending in .UnitTests.csproj
+            # There should be one [Name].UnitTests.csproj for every [Name].csproj
+            $test_projs = Get-ChildItem -Include "*.UnitTests.csproj" -Recurse
```
Comment (msetbar) [inline .circleci/continue_config.yml:85]: The test projects will have only Tests at the end not UnitTests. Do you want me to rename them to .UnitTests.csproj?

Comment (msetbar) [inline .circleci/continue_config.yml:85]: The default folder is always the location of repo?

File: .circleci/continue_config.yml (hunk 2 above)
Comment (dillon-ms) [reply .circleci/continue_config.yml:75]: @msetbar https://github.com/sightmachine/azure-integrations/pull/89 here's the pr into azure-integrations. I wrote my reasoning there in response to oscar

File: .gitignore [hunk 4]
```diff
@@ -301,5 +301,16 @@ __pycache__/
...
\ No newline at end of file
+## Mac Os
+
+.Ds_Store
+
+## Testing output
+TestResults/
+CoverageReport/
+coverage.cobertura.xml
+Tools/
```
Comment (msetbar) [inline .gitignore:312]: This will exclude tools folder.

Comment (dillon-ms) [reply .gitignore:312]: this was just from testing my POC, please look here  https://github.com/sightmachine/azure-integrations/pull/89

**Pull request closed without merging**